src/lib/markbind/src/preprocessors/componentPreprocessor.js

```diff
@@ -0,0 +1,394 @@
+
+  _deleteIncludeAttributes(element);
+
+  // Process sources with or without hash, retrieving and appending
+  // the appropriate children to a wrapped include element
+
+  let actualContent;
+
+  if (hash) {
+    // Keep scripts in the fileContent
+    const src = cheerio.parseHTML(fileContent, true);
+    const $ = cheerio.load(src);
+    const hashContent = $(hash).html();
+
+    actualContent = (hashContent && isTrim) ? hashContent.trim() : hashContent;
+
+    if (actualContent === null && isOptional) {
+      // Use empty content if it is optional
+      actualContent = '';
+    } else if (actualContent === null) {
+      const hashSrcWithoutHash = hash.substring(1);
+      const error = new Error(`No such segment '${hashSrcWithoutHash}' in file: ${actualFilePath}\n`
+          + `Missing reference in ${element.attribs[ATTRIB_CWF]}`);
+      parser._onError(error);
+
+      return utils.createErrorNode(element, error);
+    }
+
+    // optional includes of segments have now been handled, so delete the attribute
+    if (isOptional) delete element.attribs.optional;
+  } else {
+    actualContent = (fileContent && isTrim) ? fileContent.trim() : fileContent;
+  }
+
+  if (isIncludeSrcMd) {
+    if (context.mode === 'include') {
+      actualContent = isInline ? actualContent : `\n\n${actualContent}\n`;
+    } else {
+      // TODO do we need this branch?
+      actualContent = md.render(actualContent);
+    }
+
+    // TODO clarify why this is only done if there is a hash, or fix it if it is unintended
+    if (hash) _rebaseReferenceForStaticIncludes(actualContent, element, config);
+  }
+
+  // Flag with a data-included-from flag with the source filePath for calculating
+  // the file path of dynamic resources ( images, anchors, plugin sources, etc. ) later
+  const wrapperType = isInline ? 'span' : 'div';
+  const childrenHtml = `<${wrapperType} data-included-from="${filePath}">${actualContent}</${wrapperType}>`;
+  element.children = cheerio.parseHTML(childrenHtml, true);
+
+  // Recurse on children, but return an error node if the call stack has exceeded the maximum
+  if (element.children && element.children.length > 0) {
+    if (childContext.callStack.length > CyclicReferenceError.MAX_RECURSIVE_DEPTH) {
+      const error = new CyclicReferenceError(childContext.callStack);
+      parser._onError(error);
+      return utils.createErrorNode(element, error);
+    }
+
+    // eslint-disable-next-line no-use-before-define
+    element.children = element.children.map(e => preProcessComponent(e, childContext, config, parser));
+  }
+
+  return element;
+}
+
+function _preprocessVariables() {
+  return utils.createEmptyNode();
+}
+
+function _preprocessBody(node) {
+  // eslint-disable-next-line no-console
+  console.warn(`<body> tag found in ${node.attribs[ATTRIB_CWF]}. This may cause formatting errors.`);
+}
+
+function preProcessComponent(node, context, config, parser) {
```

Comment:
Entry point to the original `_preprocess` function, the respective switch cases are the respective control flows of `_preprocess` for different types of elements